### PR TITLE
Fixes for trl upgrade

### PR DESCRIPTION
**Pull request by Datta0**
For : unslothai/unsloth#2544

### CODE REVIEW

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -213,17 +271,18 @@ def grpo_accumulated_loss(
     lm_head = trainer.model.get_output_embeddings().weight
 
     with torch.amp.autocast(device_type = "cuda", dtype = mixed_dtype):
+        #breakpoint()
         with torch.inference_mode(), trainer.accelerator.unwrap_model(trainer.model, keep_fp32_wrapper = False).disable_adapter():
-            old_hidden_states = trainer.model(input_ids = input_ids, logits_to_keep = logits_to_keep + 1).logits
+            ref_hidden_states = trainer.model(input_ids = input_ids, logits_to_keep = logits_to_keep + 1).logits
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:294]: Wait so we can't do the trick of forming `XW` together for both the reference and new logits now?

File: unsloth_zoo/rl_replacements.py [hunk 2]
```diff
@@ -49,37 +49,81 @@ def selective_log_softmax(logits, index):
...
+    ref_logits,
+    new_logits,
+    old_logprobs,
+    input_ids,
+    mask,
+    beta,
+    advantages,
+    **kwargs
+):
+    # Set defaults for optional arguments
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:62]: Wait is this `torch.compile`able? and traceable? I'm surprised!

File: unsloth_zoo/rl_replacements.py [hunk 3]
```diff
@@ -49,37 +49,81 @@ def selective_log_softmax(logits, index):
...
+        
     new = new_x - torch.logsumexp(new_logits, dim = -1)
 
     # Reverse KL
-    kl_i = torch.exp(old - new) - (old - new) - 1.0
+    # Note that this is a low variance low bias estimator for the KL divergence as used in GRPO paper
+    if beta != 0.0:
+        kl_i = torch.exp(ref - new) - (ref - new) - 1.0
+    else:
+        kl_i = 0.0 # set it to 0 to not effect the downstream computation
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:95]: Very smart!

File: unsloth_zoo/rl_replacements.py [hunk 4]
```diff
@@ -49,37 +49,81 @@ def selective_log_softmax(logits, index):
...
+    # Set defaults for optional arguments
+    loss_type = kwargs.get("loss_type", "grpo")
+    epsilon_low = kwargs.get("epsilon_low", 0.2)
+    epsilon_high = kwargs.get("epsilon_high", 0.2)
+    max_completion_length = kwargs.get("max_completion_length", 8192)
+    delta = kwargs.get("delta", None)
+
     # All Unsloth Zoo code licensed under LGPLv3
-    old_logits = old_logits.to(torch.float32)
     new_logits = new_logits.to(torch.float32)
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:70]: If I remember in Unsloth's PR, an explicit upcast was done to float32 -> since we already do this inside this function, I'm assuming the Unsloth PR's upcast is not necessary? (PLEASE CHECK AND CONFIRM!)

File: unsloth_zoo/rl_replacements.py [hunk 5]
```diff
@@ -104,14 +148,21 @@ def grpo_compute_loss(old_logits, new_logits, input_ids, mask, beta, advantages)
...
+            extra_kwargs = {}
+        def compute_loss(new_hidden_states, old_logprobs, ref_hidden_states,input_ids, mask, advantages, scaling):
             new_logits = torch.matmul(new_hidden_states, lm_head.t())
             new_logits = new_logits[:, :-1, :] # exclude the last logit: it corresponds to the next token pred
-            old_logits = torch.matmul(old_hidden_states, lm_head.t())
-            old_logits = old_logits[:, :-1, :] # exclude the last logit: it corresponds to the next token pred
+            ref_logits = torch.matmul(ref_hidden_states, lm_head.t())
+            ref_logits = ref_logits[:, :-1, :] # exclude the last logit: it corresponds to the next token pred 
+            # if old_hidden_states is not None: 
+            #     old_logits = torch.matmul(old_hidden_states, lm_head.t()) #last logit already excluded
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:177]: The matmul can be left as float16

File: unsloth_zoo/rl_replacements.py [hunk 6]
```diff
@@ -145,7 +196,11 @@ def accumulate_chunk(new_hidden_states_j, old_hidden_states_j, input_ids_j, mask
 
         grad_inputs_chunks = torch.chunk(grad_inputs,        chunks = n_chunks, dim = 0)
         new_hidden_states  = torch.chunk(_new_hidden_states, chunks = n_chunks, dim = 0)
-        old_hidden_states  = torch.chunk(_old_hidden_states, chunks = n_chunks, dim = 0)
+        if _old_logprobs is not None: 
+            old_logprobs  = torch.chunk(_old_logprobs, chunks = n_chunks, dim = 0)
+        else:
```
Comment (danielhanchen) [inline unsloth_zoo/rl_replacements.py:219]: Smart! But be certain this is torch.compilable and traceable

**Review by danielhanchen — COMMENTED** [review]
Nice work guys

File: unsloth_zoo/rl_replacements.py (hunk 2 above)
Comment (Datta0) [reply unsloth_zoo/rl_replacements.py:62]: Yeah lol :)
These are constants anyway so it works ? :)
